Allow setting system primary FQDN via WebUI or XMLRPC-API

ricardoasmarques opened the pull request:
## What does this PR change?

With this PR, it will now be possible to set/get the primary FQDN of a given system (via WebUI or XMLRPC-API).

**XMLRPC-API**
- The existing `system.getNetworkForSystems` method will now return a new `fqdn` field with the primary FQDN
- A new `system.setPrimaryFqdn` method has been added to set the primary FQDN of a given system

**WEBUI**
- The primary FQDN of a given systems can ve visualized/set via `System > Details > Hardware` page.

**Schema**
- The primary FQDN is stored in a new `is_primary` column on the `rhnServerFqdn` table

## GUI diff

Before:

![Screenshot from 2021-01-25 15-12-12](https://user-images.githubusercontent.com/14297426/105724547-e117f280-5f1f-11eb-97fa-9897e5bb28b4.png)

After:

![107793387-24f75e00-6d4e-11eb-9214-378c32f62231](https://user-images.githubusercontent.com/14297426/109627265-57200280-7b39-11eb-8b5b-4c41f276c698.png)

- [x] **DONE**

## Documentation
- Documentation issue was created: https://github.com/SUSE/spacewalk/issues/13759

- [x] **DONE**

## Test coverage
- No tests: **add explanation**
- No tests: already covered
- Unit tests were added
- Cucumber tests were added

- [ ] **DONE**

## Links

Fixes https://github.com/SUSE/spacewalk/issues/13480

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)


## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql" (Test skipped, there are no changes to test)
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests" (Test skipped, there are no changes to test)
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

moio commented:
@ricardoasmarques how would we define primary from a user perspective? What makes "primary" special compared to others?

moio commented:
@ncounter can you please take a look at the before/after screenshots above and provide a suggestion about the UX aspect? I personally find it a bit odd that two similar concepts (primary network interface and primary FQDN) are displayed differently (via a dropdown vs. radio buttons). What would be best and why, or does it make sense to keep them different?

ncounter commented:
> @ncounter can you please take a look at the before/after screenshots above and provide a suggestion about the UX aspect? I personally find it a bit odd that two similar concepts (primary network interface and primary FQDN) are displayed differently (via a dropdown vs. radio buttons). What would be best and why, or does it make sense to keep them different?

From my perspective we should align the two for having the same UX, if possible. IIUC the UX is exactly the same: the user picks one among the multiple options, and the picked one is the `primary`, no more than that. A couple of picky question to choose the right graphical object to use:
1- if the user is **not** looking for changing the `primary` value, are other values useful to be displayed, or they are just taking space because the user is not looking at them unless he really wants to change it? [taking space | have value]
2- how many options do we usually expect to have? [less than 5 | more than 5]

Answer to 1 is more important than 2 actually, anyway this is the guideline:
- use a `select` (AKA `dropdown`) if the main value we need to take care is the "chosen" one, and other options can be hided. Also, use a `select` if the list of multiple options is veeeery long, because when you don't need them they are collapsed.
- use a `radio` button multiple option if the list is short (otherwise it takes too much space and it confuses the user) and also if you still want to see all of the other possibilities once you are done with the selection. I.e. when I need to know all the properties, but I need to name one as primary but still, I want to see them all.

Wrapping up: from my POV a `select` is a good element in both of the cases here. In addition to that, we should also consider to rename the label of the `FQDNs` to `Primary FQDN`, because when the dropdown is collapsed, we will see only the primary selection, and that's the one the user wants to see in the end.

@moio : do all of my assumptions make any sense or are they wrong, from what the user wants perspective?

If for some reason we instead need to present all of the values, I'd rather go with a `radio` button approach, but with something different: a list of options where you can click on the value and the one you click gets a badge next to it saying `primary`, and the others have a "make this primary" clickable text.

@ricardoasmarques I'd still go for the `select` option, but if for some reason we need to fallback on this option, ping me again I will prepare a mock-up or something (a picture explains it better than 1000 words sometime :sweat_smile: )

ricardoasmarques commented:
@ncounter @moio Thanks for the feedback, I've now changed it to a dropdown and updated the label accordingly:

![Screenshot from 2021-02-12 16-10-44](https://user-images.githubusercontent.com/14297426/107793387-24f75e00-6d4e-11eb-9214-378c32f62231.png)

moio commented on schema/spacewalk/upgrade/susemanager-schema-4.2.7-to-susemanager-schema-4.2.8/701-rhn_server_fqdn.sql:
```diff
@@ -0,0 +1,6 @@
+
+ALTER TABLE rhnServerFQDN ADD COLUMN IF NOT EXISTS is_primary VARCHAR(1);
```
Would it make sense to set one per server as the primary as part of the migration?

moio commented on java/spacewalk-java.changes:
```diff
@@ -1,3 +1,5 @@
+- Allow setting system primary FQDN via WebUI or XMLRPC-API
```
```suggestion
- Allow setting a primary FQDN per client, either via WebUI or XMLRPC-API
```

moio commented on java/code/src/com/redhat/rhn/domain/server/ServerFQDN.java:
```diff
@@ -73,6 +74,20 @@ public void setName(String nameIn) {
+    /**
+     * @return primary String which indicates primary interface
+     */
+    public String getPrimary() {
```
Could we turn those accessors into `boolean` ones? I find the "YN" string confusing.

See how this is done, for example, in the PAYG attribute of Server:

https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/domain/server/Server_legacyUser.hbm.xml#L35

https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/domain/server/Server.java#L130

https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/domain/server/Server.java#L2079-L2093

My understanding was that the "magic" `yes_no` Hibernate column type hides away all string handling - or at least that would be my hope :angel:

moio commented on java/code/src/com/redhat/rhn/domain/server/Server.java:
```diff
@@ -945,6 +947,26 @@ public String getIpAddress() {
+     * @param fqdnName The FQDN to be checked
+     * @return Returns true if the FQDN exists, otherwise false
+     */
+    public boolean existsFqdn(String fqdnName) {
```
Does this method have to be `public`?

moio commented on java/code/src/com/redhat/rhn/frontend/xmlrpc/system/SystemHandler.java:
```diff
@@ -1866,6 +1868,7 @@ public int deleteSystem(User loggedInUser, Integer serverId, String cleanupType)
             network.put("ip", StringUtils.defaultString(server.getIpAddress()));
             network.put("ip6", StringUtils.defaultString(server.getIp6Address()));
             network.put("hostname", StringUtils.defaultString(server.getHostname()));
+            network.put("fqdn", StringUtils.defaultString(server.getFqdn()));
```
Since we are adding a new attribute here, should we make clear this is the primary one?

ricardoasmarques replied:
I've renamed it from `fqdn` to `primary_fqdn`.

moio commented on java/code/src/com/redhat/rhn/frontend/xmlrpc/NoSuchFQDNException.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * No Such FQDN Interface Exception
+ *
+ * @version $Rev$
```
These tags should be removed everywhere and definitely not appear in new code :sweat_smile:

ricardoasmarques replied:
See https://github.com/uyuni-project/uyuni/pull/3294

moio commented on java/code/src/com/redhat/rhn/frontend/xmlrpc/system/SystemHandler.java:
```diff
@@ -6605,6 +6608,30 @@ public int setPrimaryInterface(User loggedInUser, Integer serverId,
+    public int setPrimaryFqdn(User loggedInUser, Integer serverId, String fqdn) {
+        Server server = lookupServer(loggedInUser, serverId);
+
+        if (!server.existsFqdn(fqdn)) {
```
Would it be more straightforward to have a finder method that returns an `Optional<ServerFQDN>` instance and set the flag there if it is found?

ricardoasmarques commented:
@moio I believe I've now addressed all your comments, can you please do another round of review? Thanks!

witekest reviewed (commented):
Tested with Uyuni SD. Works as expected. Two nits inline.
An implementation detail about storing server is not clear to me.

ricardoasmarques commented on java/code/src/com/redhat/rhn/frontend/xmlrpc/NoSuchFQDNException.java:
```diff
@@ -0,0 +1,35 @@
+    public NoSuchFQDNException(String fqdn) {
+        super(1074, "No Such FQDN" , LocalizationService.getInstance().
+                getMessage("api.system.nosuchfqdn",
+                new Object [] {fqdn}));
```
Done

ricardoasmarques commented on java/code/src/com/redhat/rhn/domain/server/Server.java:
```diff
@@ -945,6 +946,36 @@ public String getIpAddress() {
+     */
+    public ServerFQDN findPrimaryFqdn() {
+        return this.fqdns.stream()
+                .filter((fqdn) -> fqdn.isPrimary())
```
Done

ricardoasmarques commented on java/code/src/com/redhat/rhn/domain/server/Server.java:
```diff
@@ -1955,6 +1988,23 @@ private NetworkInterface lookupForPrimaryInterface() {
+                newPrimaryFQDN = fqdn;
+            }
+        }
+        SystemManager.storeServer(this);
```
This `storeServer` call is needed to guarantee that we don't hit the DB unique constraint that prevents more than one primary FQDN for each server, but we don't have to call it again because DB will then be updated via SessionFilter:

https://github.com/uyuni-project/uyuni/blob/93a51e3edc32dbc7cbaccc53d8da5743aeba9176/java/code/src/com/redhat/rhn/frontend/servlets/SessionFilter.java

witekest commented on java/code/src/com/redhat/rhn/frontend/xmlrpc/system/SystemHandler.java:
```diff
@@ -1853,6 +1855,7 @@ public int deleteSystem(User loggedInUser, Integer serverId, String cleanupType)
      *       #prop_desc("string", "ip", "IPv4 address of system")
      *       #prop_desc("string", "ip6", "IPv6 address of system")
      *       #prop_desc("string", "hostname", "Hostname of system")
+     *       #prop_desc("string", "fqdn", "FQDN of system")
```
```suggestion
     *       #prop_desc("string", "primary_fqdn", "Primary FQDN of system")
```

ricardoasmarques replied:
Nice catch! Fixed, thanks.

moio reviewed (approved):
I think the changes are good to go, the only other requirement from my part is some minimal coverage in our Cucumber testsuite, as this touched the UI.

Thanks!

ricardoasmarques commented:
A minimal Cucumber test has been added.
